packages/framework/presence/src/presenceDatastoreManager.ts

```diff
@@ -0,0 +1,316 @@
+				(audienceClientId) => audienceClientId !== clientId,
+			);
+			// Limit to three providers to prevent flooding the network.
+			// If none respond, others present will (should) after a delay.
+			if (updateProviders.length > 3) {
+				updateProviders.length = 3;
+			}
+			this.runtime.submitSignal(joinMessageType, {
+				sendTimestamp: Date.now(),
+				avgLatency: this.averageLatency,
+				data: this.datastore,
+				updateProviders,
+			} satisfies ClientJoinMessage["content"]);
+		});
+		runtime.on("signal", this.processSignal.bind(this));
+	}
+
+	public getWorkspace<TSchema extends PresenceStatesSchema>(
+		internalWorkspaceAddress: string,
+		requestedContent: TSchema,
+	): PresenceStates<TSchema> {
+		const existing = this.workspaces.get(internalWorkspaceAddress);
+		if (existing) {
+			return existing.internal.ensureContent(requestedContent);
+		}
+
+		let workspaceDatastore = this.datastore[internalWorkspaceAddress];
+		if (workspaceDatastore === undefined) {
+			workspaceDatastore = this.datastore[internalWorkspaceAddress] = {};
+		}
+
+		const localUpdate = (
+			stateKey: string,
+			value: ClientUpdateEntry,
+			forceBroadcast: boolean,
+		): void => {
+			const clientId = this.runtime.clientId;
+			if (clientId === undefined) {
+				return;
+			}
+
+			this.localUpdate(
+				{
+					[internalWorkspaceAddress]: {
+						[stateKey]: { [clientId]: value },
+					},
+				},
+				forceBroadcast,
+			);
+		};
+
+		const entry = createPresenceStates(
+			{
+				clientId: () => this.runtime.clientId,
+				lookupClient: this.presence.getAttendee.bind(this.presence),
+				localUpdate,
+			},
+			workspaceDatastore,
+			requestedContent,
+		);
+
+		this.workspaces.set(internalWorkspaceAddress, entry);
+		return entry.public;
+	}
+
+	private localUpdate(
+		data: GeneralDatastoreMessageContent & Partial<SystemDatastore>,
+		_forceBroadcast: boolean,
+	): void {
+		const content = {
+			sendTimestamp: Date.now(),
+			avgLatency: this.averageLatency,
+			// isComplete: false,
+			data,
+		} satisfies DatastoreUpdateMessage["content"];
+		this.runtime.submitSignal(datastoreUpdateMessageType, content);
+	}
+
+	private broadcastAllKnownState(): void {
+		this.runtime.submitSignal(datastoreUpdateMessageType, {
+			sendTimestamp: Date.now(),
+			avgLatency: this.averageLatency,
+			isComplete: true,
+			data: this.datastore,
+		} satisfies DatastoreUpdateMessage["content"]);
+		this.refreshBroadcastRequested = false;
+	}
+
+	public processSignal(
+		message: IInboundSignalMessage | DatastoreUpdateMessage | ClientJoinMessage,
+		local: boolean,
+	): void {
+		const received = Date.now();
+		assert(message.clientId !== null, "Map received signal without clientId");
+		if (!isPresenceMessage(message)) {
+			return;
+		}
+		if (local) {
+			const deliveryDelta = received - message.content.sendTimestamp;
+			this.returnedMessages = Math.min(this.returnedMessages + 1, 256);
```

Comment:
Recommend either commenting this 256 (and any logic behind its choice) or name it as a const.

Based on a quick read, I'm not clear if this latency-averaging calculation is just for telemetry or if you're intending to use it functionally.  I'm hoping just telemetry?

Reply:
Comments pending.
Nah, it is a requested feature. In the spec, but very experimental. If it looks like it could be viable, we'll likely make some adjustments in services for results. Holding back on requests for now.